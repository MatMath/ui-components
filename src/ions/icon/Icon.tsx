import styled from 'styled-components';
import { Props as IconProps } from '@atoms/svg/base/SvgIcon';

export type ColorProps = 'color' | 'hoverColor' | 'activeColor';
interface Props extends Omit<IconProps, 'title' | 'children'> {
  animated?: boolean;
}

const getSize = ({ size = 20 }) => `${size}px`;
const getColor = (propName: ColorProps = 'color') => (props: Props): string =>
  props.animated ? 'none' : props[propName] || props['color'] || 'inherit';

const Icon = styled.svg.attrs<Props>(({ viewBox }: Props) => ({
  viewBox: viewBox || '0 0 20 20',
  version: '1.1',
  xmlns: 'http://www.w3.org/2000/svg'
}))`
  display: flex;
  width: ${getSize};
  height: ${getSize};
  cursor: ${({ disabled }): string => (disabled ? 'not-allowed' : 'pointer')};
  outline: none;
  border-radius: 4px;
  fill: ${getColor()};
  opacity: ${({ disabled }): number => (disabled ? 0.5 : 1)};
  transition: ${({ animated }): string =>
    animated ? 'none' : 'all 0.1s ease-in'};
  &:hover {
    fill: ${getColor('hoverColor')};
  }
  &:active {
    fill: ${getColor('activeColor')};
  }
`;

export default Icon;
