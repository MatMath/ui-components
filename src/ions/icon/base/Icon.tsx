import styled from 'styled-components';
import { Props as IconProps } from '@ions/icon/types';
import { dashGreen00 } from '@colors';

export type ColorProps = 'color' | 'hoverColor' | 'activeColor';
interface Props extends Omit<IconProps, 'title' | 'children'> {
  animated?: boolean;
}

const getWidth = ({ width, size }: Props) => `${size || width || 20}px`;
const getHeight = ({ height, size }: Props) => `${size || height || 20}px`;
const getColor = (propName: ColorProps = 'color') => (props: Props): string =>
  props.animated ? 'none' : props[propName] || props['color'] || dashGreen00;

export const Icon = styled.svg.attrs<Props>(({ viewBox }: Props) => ({
  viewBox: viewBox || '0 0 20 20',
  version: '1.1',
  xmlns: 'http://www.w3.org/2000/svg'
}))`
  display: flex;
  width: ${getWidth};
  height: ${getHeight};
  cursor: ${({ disabled }): string => (disabled ? 'not-allowed' : 'inherit')};
  outline: none;
  border-radius: 4px;
  fill: ${getColor()};
  opacity: ${({ disabled }): number => (disabled ? 0.5 : 1)};
  transition: ${({ animated }): string =>
    animated ? 'none' : 'all 0.1s ease-in'};
  &:hover {
    fill: ${getColor('hoverColor')};
  }
  &:active {
    fill: ${getColor('activeColor')};
  }
`;
