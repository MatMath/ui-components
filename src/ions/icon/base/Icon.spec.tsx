import * as React from 'react';
import { mount } from 'enzyme';
import 'jest-styled-components';
import Icon from './Icon';
import { dashGreen00 } from '@colors';

describe('<Icon>', () => {
  describe('global render', () => {
    it('should render', () => {
      const IconWrapper = mount(<Icon />);

      expect(IconWrapper).toMatchSnapshot();
    });

    describe('component style', () => {
      it('should set a default style if none is provided', () => {
        const IconWrapper = mount(<Icon />);

        expect(IconWrapper).toHaveStyleRule('fill', dashGreen00);
        expect(IconWrapper).toHaveStyleRule('cursor', 'inherit');
        expect(IconWrapper).toHaveStyleRule('width', '20px');
        expect(IconWrapper).toHaveStyleRule('height', '20px');
      });

      it('should set the provided style', () => {
        const IconWrapper = mount(<Icon size={40} color='red' />);

        expect(IconWrapper).toHaveStyleRule('fill', 'red');
        expect(IconWrapper).toHaveStyleRule('width', '40px');
        expect(IconWrapper).toHaveStyleRule('height', '40px');
      });

      it('should handle disabled mode', () => {
        const IconWrapper = mount(<Icon disabled />);

        expect(IconWrapper).toHaveStyleRule('opacity', '0.5');
        expect(IconWrapper).toHaveStyleRule('cursor', 'not-allowed');
      });
    });
  });
});
