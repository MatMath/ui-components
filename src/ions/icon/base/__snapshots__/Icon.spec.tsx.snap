// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon> global render should render 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 20px;
  height: 20px;
  cursor: inherit;
  outline: none;
  border-radius: 4px;
  fill: #0E353D;
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
  opacity: 1;
  -webkit-transition: all 0.1s ease-in;
  transition: all 0.1s ease-in;
}

.c0:hover {
  fill: #0E353D;
}

.c0:active {
  fill: #0E353D;
}

<styled.svg>
  <StyledComponent
    forwardedComponent={
      Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [
          [Function],
        ],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bdVaJa",
          "isStatic": false,
          "lastClassName": "c0",
          "rules": Array [
            "
  display: flex;
  width: ",
            [Function],
            ";
  height: ",
            [Function],
            ";
  cursor: ",
            [Function],
            ";
  outline: none;
  border-radius: 4px;
  fill: ",
            [Function],
            ";
  transform: ",
            [Function],
            ";
  opacity: ",
            [Function],
            ";
  transition: ",
            [Function],
            "
  &:hover {
    fill: ",
            [Function],
            ";
  }
  &:active {
    fill: ",
            [Function],
            ";
  }
",
          ],
        },
        "displayName": "styled.svg",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-bdVaJa",
        "target": "svg",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      }
    }
    forwardedRef={null}
  >
    <svg
      className="c0"
      version="1.1"
      viewBox="0 0 20 20"
      xmlns="http://www.w3.org/2000/svg"
    />
  </StyledComponent>
</styled.svg>
`;
